packages/flutter_driver/lib/src/common/wait.dart

```diff
@@ -0,0 +1,185 @@
+
+  @override
+  Future<void> wait() async {
+    if (condition) {
+      return;
+    } else {
+      do {
+        await SchedulerBinding.instance.endOfFrame;
+      } while (!condition);
+    }
+    assert(condition);
+  }
+
+  @override
+  Map<String, String> serialize() => <String, String>{
+        'conditionName': 'NoTransientCallbacksCondition',
+      };
+}
+
+/// A condition that waits until no pending frame is scheduled.
+class NoPendingFrameCondition implements WaitCondition {
+  /// Creates a [NoPendingFrameCondition] instance.
+  const NoPendingFrameCondition();
+
+  /// Factory constructor to parse a [NoPendingFrameCondition] instance from the
+  /// given JSON map.
+  factory NoPendingFrameCondition.fromJson(Map<String, dynamic> json) {
+    if (json['conditionName'] != 'NoPendingFrameCondition')
+      throw SerializationException(
+          'Not a NoPendingFrameCondition json string.');
+    return const NoPendingFrameCondition();
+  }
+
+  @override
+  bool get condition => !SchedulerBinding.instance.hasScheduledFrame;
+
+  @override
+  Future<void> wait() async {
+    if (condition) {
+      return;
+    } else {
+      do {
+        await SchedulerBinding.instance.endOfFrame;
+      } while (!condition);
+    }
+    assert(condition);
+  }
+
+  @override
+  Map<String, String> serialize() => <String, String>{
+        'conditionName': 'NoPendingFrameCondition',
+      };
+}
+
+/// A combined condition that waits until all the given [conditions] are met.
+class CombinedCondition implements WaitCondition {
+  /// Creates a [CombinedCondition] instance with the given list of
+  /// [conditions].
+  const CombinedCondition(this.conditions);
+
+  /// Factory constructor to parse a [CombinedCondition] instance from the given
+  /// JSON map.
+  factory CombinedCondition.fromJson(Map<String, dynamic> jsonMap) {
+    if (jsonMap['conditionName'] != 'CombinedCondition')
+      throw SerializationException('Not a CombinedCondition json string.');
+    CombinedCondition combinedCondition;
+    if (jsonMap['conditions'] == null) {
+      combinedCondition = const CombinedCondition(<WaitCondition>[]);
+    } else {
+      final List<WaitCondition> conditions = <WaitCondition>[];
+      for (Map<String, dynamic> condition
+          in json.decode(jsonMap['conditions'])) {
+        conditions.add(WaitConditionDecoder.fromJson(condition));
+      }
+      combinedCondition = CombinedCondition(conditions);
+    }
+    return combinedCondition;
+  }
+
+  /// A list of conditions it waits for.
+  final List<WaitCondition> conditions;
+
+  @override
+  bool get condition =>
+      conditions.every((WaitCondition condition) => condition.condition);
+
+  @override
+  Future<void> wait() async {
+    do {
+      for (WaitCondition condition in conditions) {
+        if (condition != null) {
+          await condition.wait();
+        }
+      }
+    } while (!condition);
```

Comment:
why not a while instead of a do/while? Why await at all if things are already true?

Reply:
Done.